cmd/run.go

```diff
@@ -290,36 +291,18 @@ a commandline interface for interacting with it.`,
 				logger.Warn("No script iterations finished, consider making the test duration longer")
 			}
 
-			data := ui.SummaryData{
-				Metrics:   engine.Metrics,
-				RootGroup: engine.ExecutionScheduler.GetRunner().GetDefaultGroup(),
-				Time:      executionState.GetCurrentTestRunDuration(),
-				TimeUnit:  conf.Options.SummaryTimeUnit.String,
-			}
-			// Print the end-of-test summary.
+			// Handle the end-of-test summary.
 			if !runtimeOptions.NoSummary.Bool {
-				fprintf(stdout, "\n")
-
-				s := ui.NewSummary(conf.SummaryTrendStats)
-				s.SummarizeMetrics(stdout, "", data)
-
-				fprintf(stdout, "\n")
-			}
-
-			if runtimeOptions.SummaryExport.ValueOrZero() != "" { //nolint:nestif
-				f, err := os.Create(runtimeOptions.SummaryExport.String)
+				summaryResult, err := initRunner.HandleSummary(globalCtx, &lib.Summary{
+					Metrics:         engine.Metrics,
+					RootGroup:       engine.ExecutionScheduler.GetRunner().GetDefaultGroup(),
+					TestRunDuration: executionState.GetCurrentTestRunDuration(),
+				})
+				if err == nil {
+					err = handleSummaryResult(afero.NewOsFs(), os.Stdout, os.Stderr, summaryResult)
+				}
 				if err != nil {
-					logger.WithError(err).Error("failed to create summary export file")
-				} else {
-					defer func() {
-						if err := f.Close(); err != nil {
-							logger.WithError(err).Error("failed to close summary export file")
-						}
-					}()
-					s := ui.NewSummary(conf.SummaryTrendStats)
-					if err := s.SummarizeMetricsJSON(f, data); err != nil {
-						logger.WithError(err).Error("failed to make summary export file")
-					}
+					logger.WithError(err).Error("failed to handle the end-of-test summary")
```

Comment:
I guess that in a sequential release we will fallback to a default representation or the dumping the `lib.Summary` as JSON to the disk?

Reply:
not sure what you mean by "a sequential release", but we'll get an error here only if something goes majorly wrong... Both `initRunner.HandleSummary()` and `handleSummaryResult()` try to continue working even if there are errors. `handleSummaryResult()` tries to save all files, even if some of them fail (because of wrong names, lack of permissions, etc.). And `initRunner.HandleSummary()` has internal `try / catch` in the JS wrapper code: https://github.com/loadimpact/k6/blob/0838813db6edead0f6faa3cdaadf3b5eb070a1ab/js/summary.go#L97-L104

Though I should add a test for that...

Reply:
yeah, I missed this code at the time ...

This is probably my biggest issue reviewing the PR, the code is all over the place, although I don't know how to fix that if it's at all possible given what the code tries to do, so I don't have some concrete feedback. I hope that when we rewrite even more of the whole metric and threshold (which specifically are also over the place) this might get more streamlined